### PR TITLE
Timed sorting with sortWithIn operation #15 for review

### LINKED ISSUE #15: streams: request new feature: timed sorting
Clone of https://github.com/akka/akka/issues/20643

Can we have timed sorting operation in akka streams?

Similar to `takeWithin`, This could be `sortWithin`

Possible method signatures
```scala
def sortWithin(d: FiniteDuration): Source[Out, Mat]
def sortWithin(d: FiniteDuration, c: Comparator): Source[Out, Mat]
```

### PR DESCRIPTION
**Pull request by anandshah123**
Timed sorting with sortWithIn operation (Closes #15)

### CODE REVIEW

**Comment by typesafehub-validator** [comment]
Hi @anandshah123,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Typesafe Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

File: src/main/scala/akka/stream/contrib/SortWithin.scala [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import akka.stream.impl.fusing.GraphStages.SimpleLinearGraphStage
+import akka.stream.stage.{GraphStageLogic, InHandler, OutHandler, TimerGraphStageLogic}
+
+import scala.collection.immutable
+import scala.concurrent.duration.FiniteDuration
+
+object SortWithin {
+
+  /**
+    * Factory for [[SortWithin]] instances.
```
Comment (ktoso) [inline src/main/scala/akka/stream/contrib/SortWithin.scala]: Please use the usual formatting (as Akka does):

```
/**
   * Factory for [[SortWithin]] instances.
```

File: src/main/scala/akka/stream/contrib/SortWithin.scala [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    *
+    * @param finiteDuration duration within which sorting needs to be done. Ideally should be kept as small as possible
+    * @tparam A input type
+    * @return [[SortWithin]] instance.
```
Comment (ktoso) [inline src/main/scala/akka/stream/contrib/SortWithin.scala]: no need to document the obvious things, please remove

File: src/main/scala/akka/stream/contrib/SortWithin.scala [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    */
+  def apply[A <: Comparable[A]](finiteDuration: FiniteDuration): SortWithin[A] = new SortWithin[A](finiteDuration)
+
+  /**
+    * Java API: Factory for [[SortWithin]] instances.
```
Comment (ktoso) [inline src/main/scala/akka/stream/contrib/SortWithin.scala]: good

Comment (ktoso) [inline src/main/scala/akka/stream/contrib/SortWithin.scala]: There should be documentation what SoprtWithin actually does, not that we're a factory here :)

File: src/main/scala/akka/stream/contrib/SortWithin.scala [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    *
+    * @param finiteDuration duration within which sorting needs to be done. Ideally should be kept as small as possible
+    * @tparam A input type
+    * @return [[SortWithin]] instance.
+    */
+  def create[A <: Comparable[A]](finiteDuration: FiniteDuration): SortWithin[A] = new SortWithin[A](finiteDuration)
+}
+
+final class SortWithin[T <: Comparable[T]](timeout: FiniteDuration) extends SimpleLinearGraphStage[T] {
```
Comment (ktoso) [inline src/main/scala/akka/stream/contrib/SortWithin.scala]: what if `0.millis`?

File: src/main/scala/akka/stream/contrib/SortWithin.scala [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        pull(in)
+      }
+
+      @scala.throws[Exception](classOf[Exception])
+      override def onUpstreamFinish(): Unit = {
+        emitMultiple(out, sortedElements.iterator)
+        super.onUpstreamFinish()
+      }
+
+      @scala.throws[Exception](classOf[Exception])
```
Comment (ktoso) [inline src/main/scala/akka/stream/contrib/SortWithin.scala]: no need to include these annotations, please remove

File: src/test/scala/akka/stream/contrib/SortWithinSpec.scala [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+      source.sendNext(1)
+      source.sendNext(3)
+      source.sendNext(2)
+      source.sendComplete()
+      sink.expectNext(1, 2, 3)
+      sink.expectComplete()
+    }
+  }
+
+  "A stream via SortWithin" should {
```
Comment (ktoso) [inline src/test/scala/akka/stream/contrib/SortWithinSpec.scala]: please nest the 3 tests in the same `"A stream via SortWithin" should {` block :)

File: src/test/scala/akka/stream/contrib/SortWithinSpec.scala [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        .toMat(TestSink.probe)(Keep.both)
+        .run()
+      sink.request(99)
+      source.sendNext(3)
+      source.sendNext(1)
+      val cause: RuntimeException = new RuntimeException
+      source.sendError(cause)
+      sink.expectNext(FiniteDuration.apply(3, TimeUnit.SECONDS), 1)
+      sink.expectError(cause)
+    }
```
Comment (ktoso) [inline src/test/scala/akka/stream/contrib/SortWithinSpec.scala]: seems to be missing a test for the case when we emit 3 elements, timer triggers, then we emit some more elements, they should be independently sorted

Comment (ktoso) [inline src/test/scala/akka/stream/contrib/SortWithinSpec.scala]: Hm I see this is doing something a bit different...
May I ask on the exact use case?

I was expecting this to "sort as many elements as you can within T time, emit it as sorted sequence".
Why is this like a "take within, but sorted"? If that's the goal one can simply takeWithin and then sort once.

**Comment by ktoso** [comment]
Thanks for the contrib, had some questions in line, most interestingly for me is the actual use case. 
This doesn't seem to do more than takeWithin+sort though, or am I missing something?

File: src/main/scala/akka/stream/contrib/SortWithin.scala (hunk 4 above)
Comment (anandshah123) [inline src/main/scala/akka/stream/contrib/SortWithin.scala]: Added validation for the same

File: src/test/scala/akka/stream/contrib/SortWithinSpec.scala (hunk 7 above)
Comment (anandshah123) [inline src/test/scala/akka/stream/contrib/SortWithinSpec.scala]: Yes you are correct use case is "sort as many elements possible within T time, emit sorted sequence & repeat"

**Comment by anandshah123** [comment]
@ktoso ready for review again

File: src/main/scala/akka/stream/contrib/SortWithin.scala [hunk 8]
```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.contrib
+
+import akka.stream.Attributes
+import akka.stream.impl.fusing.GraphStages.SimpleLinearGraphStage
```
Comment (drewhk) [inline src/main/scala/akka/stream/contrib/SortWithin.scala]: Please don't use internal classes, use an ordinary GraphStage. (or alternatively, add a LinearGraphStage helper to contrib itself that is no longer internal)

File: src/main/scala/akka/stream/contrib/SortWithin.scala (hunk 4 above)
Comment (drewhk) [inline src/main/scala/akka/stream/contrib/SortWithin.scala]: Does this logic need a GraphStage at all? Can't you just use:

```scala
Flow[A].groupedWithin(duration).mapConcat(identity)
```

**Comment by anandshah123** [comment]
I don't think tarvis build should fail. Can anyone please check?

**Comment by anandshah123** [comment]
@drewhk ready for review

File: src/main/scala/akka/stream/contrib/SortWithin.scala [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    })
+
+
+    final override protected def onTimer(key: Any): Unit = {
+      emitMultiple(out, buf.result().sorted)
+      buf.clear()
+    }
+
+    override def preStart(): Unit = {
+      schedulePeriodically("SortWithinTimer", timeout)
```
Comment (patriknw) [inline src/main/scala/akka/stream/contrib/SortWithin.scala]: `private case object SortWithinTimer` in the companion would be better than string key

File: src/test/scala/akka/stream/contrib/SortWithinSpec.scala [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    "sort the elements published till given FiniteDuration" in {
+      val (source, sink) = TestSource.probe[Integer]
+        .via(SortWithin[Integer](FiniteDuration.apply(5, TimeUnit.SECONDS)))
+        .toMat(TestSink.probe)(Keep.both)
+        .run()
+      sink.request(99)
+      source.sendNext(1)
+      source.sendNext(3)
+      source.sendNext(2)
+      sink.expectNext(FiniteDuration.apply(6, TimeUnit.SECONDS), 1)
```
Comment (patriknw) [inline src/test/scala/akka/stream/contrib/SortWithinSpec.scala]: `import scala.concurrent.duration._`

and then `6.seconds`

File: src/test/scala/akka/stream/contrib/SortWithinSpec.scala [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+      sink.expectNext(FiniteDuration.apply(6, TimeUnit.SECONDS), 2)
+      sink.expectNext(FiniteDuration.apply(6, TimeUnit.SECONDS), 3)
```
Comment (patriknw) [inline src/test/scala/akka/stream/contrib/SortWithinSpec.scala]: can we speed up this test somewhat?

**Comment by anandshah123** [comment]
@patriknw ready for review

File: src/test/scala/akka/stream/contrib/SortWithinSpec.scala [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+class SortWithinSpecAutoFusingOff extends {
+  val autoFusing = false
+} with SortWithinSpec
+
+trait SortWithinSpec extends BaseStreamSpec {
+
+  "A stream via SortWithin" should {
+    "sort the elements published till given FiniteDuration" in {
+      val (source, sink) = TestSource.probe[Integer]
+        .via(SortWithin[Integer](500.millis))
```
Comment (patriknw) [inline src/test/scala/akka/stream/contrib/SortWithinSpec.scala]: This is a difficult balance, we are running CI on Travis, and that is very unpredictable with timing. Expect that you can have a GC (or other pause) for 2 seconds at any time. Perhaps something in the middle of this and your original values?

**Comment by anandshah123** [comment]
@patriknw done

**Comment by patriknw** [comment]
@anandshah123 did you answer @ktoso and @drewhk questions about why it must be a `GraphStage` instead of using `takeWithin` or `groupWithin`?

**Comment by anandshah123** [comment]
@patriknw Yes. because the intention of having this operation in place is to pass sorted elements to downstream within continuous repeating intervals. Whereas takeWithin & groupWithin are terminal operations.

**Comment by patriknw** [comment]
@drewhk @ktoso are you fine with merging this?

**Comment by drewhk** [comment]
Hm, groupedWithin is not a terminal operation, it will repeatedly emit groups sampled by time window (and possibly capped to a maximum size).

**Comment by anandshah123** [comment]
@drewhk sortWithin will sort elements & emit them to downstream by time window

**Comment by drewhk** [comment]
I still don't understand how is it different from `groupedWithin().mapConcat(identity)`

**Comment by drewhk** [comment]
To quote the docs of `groupedWithin()`:

> Chunk up this stream into groups of elements received within a time window, or limited by the given number of elements, whatever happens first. Empty groups will not be emitted if no elements are received from upstream. The last group before end-of-stream will contain the buffered elements since the previously emitted group.

**Comment by anandshah123** [comment]
@drewhk Yes `groupedWithin().mapConcat(sortFunction)` will produce same output. `sortWithin` will be like shorthand doing the same.

**Pull request closed without merging**